api/api/validators/generate_teams_validator.py

```diff
@@ -44,32 +50,167 @@ def validate_algorithm_options(self):
         pass
 
+    def _validate_student_ids_unique(self, students: List[Dict]):
+        student_ids = set()
+        for student in students:
+            student_id = student.get("id")
+            if student_id in student_ids:
+                raise SchemaError(
+                    f"Student ids must be unique. Student {student_id} is duplicated."
+                )
+            student_ids.add(student_id)
+
+    def _validate_student_project_preferences_exist(
+            self, students: List[Dict], teams: List[Dict]
+    ):
+        all_projects = set([team.get("project_id") for team in teams])
+        for student in students:
+            student_project_preferences = student.get("project_preferences")
+            if student_project_preferences is None:
+                continue
+
+            # Validate if projects under student.project_preferences exist
+            if not all_projects.issuperset(student_project_preferences):
+                raise SchemaError(
+                    f"Student {student.get('id')} has project preferences that do not exist in the project set."
+                )
+
+    def _validate_student_project_preferences(
+            self, students: List[Dict], max_project_preferences: int
+    ):
+        for student in students:
+            student_project_preferences = student.get("project_preferences")
+            if (
+                    student_project_preferences is not None
+                    and len(student_project_preferences) > max_project_preferences
+            ):
+                raise SchemaError(
+                    f"Student {student.get('id')} has {student_project_preferences} project preferences, "
+                    + f"but the maximum is {max_project_preferences}."
+                )
+
+    def _validate_student_attributes(self, students: List[Dict], teams: List[Dict]):
+        all_attributes = set()
+        for team in teams:
+            all_attributes.update(
+                [
+                    requirement.get("attribute")
+                    for requirement in team.get("requirements")
+                ]
+                if team.get("requirements")
+                else []
+            )
+        for student in students:
+            attributes = student.get("attributes")
+            if attributes is None:
+                raise SchemaError(f"Student {student.get('id')} has no attributes.")
+
+            # Validate if attribute keys integer string
+            try:
+                attribute_keys = list(map(int, attributes.keys())) if attributes else []
+            except ValueError:
+                raise SchemaError("Attribute keys must be integers.")
+
+            # Validate if attribute keys are unique
+            if len(attribute_keys) != len(set(attribute_keys)):
+                raise SchemaError("Attribute keys must be unique.")
+
+            # Validate if attribute keys exist
+            if not all_attributes.issuperset(attribute_keys):
```

Comment:
On second thought here too, it's cool if students have more attribute keys than the project requirements specify, no logical issue there. It's only minorly an issue if project ids have requirements for attributes that no student has, but I don't think we need to check that. It causes no problems for our system, just a wonky team generation for the user (i.e. user error)

tldr; remove